src/shogun/mathematics/linalg/LinalgNamespaceMatrix.cpp

```diff
@@ -0,0 +1,43 @@
+#include <shogun/lib/DataType.h>
+#include <shogun/mathematics/linalg/LinalgNamespaceMatrix.h>
+#include <shogun/mathematics/linalg/LinalgNamespace.h>
+
+#define DEFINE_SWITCH(PTYPE, FUNC, ...) \
+switch (PTYPE) \
+{ \
+	case PT_FLOAT32: \
+		FUNC<float32_t>(__VA_ARGS__); \
```

Comment:
ok so here's my thing about the former story: lazy-evaluation (see eigen) or codegen/JIT (see XLA of TF). these are not orthogonal, meaning if we can have lazy evaluation, say:
```
Matrix x, y;
float64_t i;
x.add(y).scale(i);
```
i guess you get what i'm saying... this way you only need once to figure out the type of the matrix (and other elements' type) and then the rest is a straight line. of course this complicates things a lot, but to be honest if we look around ourselves this is the flow of every framework around (a bit of functional approach in the imperative world)